Write how-to docs for token standard usage (#1872)
hyperledger-labs/splice: docs/src/app_dev/daml_api/token_standard.rst

..
   Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
..
   SPDX-License-Identifier: Apache-2.0

.. _token_standard_usage:

Wallet Integration with Token Standard Assets
=============================================

This page provides wallet developers with guidance on how to integrate with token standard assets.
Such an integration works by sending the right read and write requests to the Ledger API of the validator node hosting the wallet user's party.

> There are three kinds of integration patterns:

Reviewer: ```suggestion
There are four kinds of integration patterns:
```

  * :ref:`token_standard_usage_reading_contracts`
  * :ref:`token_standard_usage_reading_tx_history`
  * :ref:`token_standard_usage_executing_factory_choice`

>   * :ref:`token_standard_usage_executing_factory_choice`

Reviewer: non-factory choice

> All of these integration patterns are demonstrated in the form of executable code as part of the experimental command-line interface for token standard assets.

Reviewer: add link to CLI README

The sections below explaining the patterns below thus all start with a link to the code.
They then provide additional context for an implementor.

> You can find the `ledger's OpenAPI definition <https://github.com/digital-asset/canton/blob/f608ec2cbb7b3e9331b7cc564eb260916606d815/community/ledger/ledger-json-api/src/test/resources/json-api-docs/openapi.yaml#L1#L1>`_.

Reviewer: ```suggestion
All interaction works via the JSON Ledger API (see its `OpenAPI definition <https://github.com/digital-asset/canton/blob/f608ec2cbb7b3e9331b7cc564eb260916606d815/community/ledger/ledger-json-api/src/test/resources/json-api-docs/openapi.yaml#L1#L1> here`_).
```

> This is also accessible at ``http(s)://${YOUR_PARTICIPANT}/docs/openapi``.

Reviewer: ```suggestion
This OpenAPI definition is also accessible at ``http(s)://${YOUR_PARTICIPANT}/docs/openapi``.
```

We encourage developers to use OpenAPI code generation tools as opposed to manually writing HTTP requests.

> Check out the :ref:`Authentication section <app-auth>` for more information on how to authenticate the requests.

Reviewer: Why are we referring to this here? Does Ledger API auth work exactly the same as authentication to the splice apps, which is explained on that page?

If yes, then can you adjust that page to talk about both APIs?

.. _token_standard_usage_workflows:

> Token Standard Workflows
> ------------------------
> 
> There are three workflows enabled by the Token Standard.
> Refer to the CIP and source code listed here:
> 
>   * `Portfolio View <https://github.com/global-synchronizer-foundation/cips/blob/main/cip-0056/cip-0056.md#wallet-client--portfolio-view>`_
>   * `Direct Peer-to-Peer / Free of Payment (FOP) Transfers <https://github.com/global-synchronizer-foundation/cips/blob/main/cip-0056/cip-0056.md#direct-peer-to-peer--free-of-payment-fop-transfer-workflow>`_
>   * `Delivery versus Payment (DVP) Transfer Workflows <https://github.com/global-synchronizer-foundation/cips/blob/main/cip-0056/cip-0056.md#delivery-versus-payment-dvp-transfer-workflows>`_
> 

Reviewer: ```suggestion

```
at this point the easiest is to drop them. Given that more docs are going to be written by the wallet team.

The ``activeAtOffset`` parameter can be set to the result of
`the ledger-end endpoint on the participant <https://github.com/digital-asset/canton/blob/f608ec2cbb7b3e9331b7cc564eb260916606d815/community/ledger/ledger-json-api/src/test/resources/json-api-docs/openapi.yaml#L1#L711>`_
to get the latest ACS, or an older (non-pruned) one to get the ACS at that point in time.

To filter for a particular party and interface, it should include a ``filtersByParty`` with an ``InterfaceFilter``:

.. code-block:: json

  {
    "filtersByParty": {
      "$A_PARTY": {
        "cumulative": [
          {
            "identifierFilter": {
              "InterfaceFilter": {
                "value": {
                  "interfaceId": "$AN_INTERFACE_ID",
                  "includeInterfaceView": true,
                  "includeCreatedEventBlob": true
                }
              }
            }
          }
        ]
      }
    }
  }

For example:

* ``"$A_PARTY"`` could look like ``test::1220a0db3761b3fc919b55e7ff80ad740824336010bfde8829611c0e64477ab7bee5``.

> * ``"$AN_INTERFACE_ID"`` could be ``#splice-api-token-holding-v1:Splice.Api.Token.HoldingV1:Holding``.

Reviewer: ```suggestion
* ``"$AN_INTERFACE_ID"`` could be ``#splice-api-token-holding-v1:Splice.Api.Token.HoldingV1:Holding`` to read all `Holding` contracts of the specified party.
```

Additionally, there's three flags that can be set:

* ``includeInterfaceView``: to include the interface view of the contract in the response.
* ``includeCreatedEventBlob``: to include a binary blob that is required for `explicit disclosure <https://docs.daml.com/app-dev/explicit-contract-disclosure.html>`_.
* ``verbose``: to include additional information in the response.

The response for such a query will contain the ``createdEvent`` of the contract, including the interface views requested (if any).
The ``viewValue`` within it will be the JSON-serialized Daml interface view.
If more than one interface is requested, you can distinguish them by checking the ``interfaceId`` field.
You can find an `example response for Holdings here <https://github.com/hyperledger-labs/splice/blob/main/token-standard/cli/__tests__/mocks/data/holdings.json>`_.

.. _token_standard_usage_reading_tx_history:

Reading and parsing transaction history involving Token Standard contracts
--------------------------------------------------------------------------

> `Token Standard CLI's code to list transactions <https://github.com/hyperledger-labs/splice/blob/main/token-standard/cli/src/commands/listHoldingTransactions.ts>`_

Reviewer: ```suggestion
Example code: `Token Standard CLI's code to list transactions <https://github.com/hyperledger-labs/splice/blob/main/token-standard/cli/src/commands/listHoldingTransactions.ts>`_
```

                    "includeInterfaceView": true,
                    "includeCreatedEventBlob": true
                  }
                }
              }
            }
          ]
        }
      }
    }

For example:

* ``"$A_PARTY"`` could look like ``test::1220a0db3761b3fc919b55e7ff80ad740824336010bfde8829611c0e64477ab7bee5``.
* ``"$AN_INTERFACE_ID"`` could be ``#splice-api-token-holding-v1:Splice.Api.Token.HoldingV1:Holding``.

To include other transaction nodes that don't directly involve the interfaces (e.g., non-interface-specific children nodes),
a ``WildcardFilter`` can be included in the ``cumulative`` filter array:

.. code-block:: json

    {
      "identifierFilter": {
        "WildcardFilter": {
          "value": {
            "includeCreatedEventBlob": true
          }
        }
      }
    }

The ``beginExclusive`` field is the offset from which to start reading transactions.
To paginate, you can start with the ``participantPrunedUpToInclusive`` from ``GET ${PARTICIPANT_URL}/v2/state/latest-pruned-offsets``
and continue by passing the offset of the last transaction from the previous response.

Parsing the history
^^^^^^^^^^^^^^^^^^^

> You can find an example parser `here <https://github.com/hyperledger-labs/splice/blob/main/token-standard/cli/src/txparse/parser.ts>`_.

Reviewer: ```suggestion
Example code: `the parser here <https://github.com/hyperledger-labs/splice/blob/main/token-standard/cli/src/txparse/parser.ts>`_.
```

> This handles transactions involving the ``Holding`` and ``TransferInstruction`` interfaces.

Reviewer: ```suggestion
extracts a user-readable wallet history by parsing transactions involving the ``Holding`` and ``TransferInstruction`` interfaces.
```

Where further customization is required, a parser can decide to also focus on internal/specific choices that are not available in the standard, but in some specific implementation.

In each Token Standard exercise node, one can find:

* The choice being executed, useful to distinguish what operation was performed.
* As part of the archival/creation of children, one can find out other relevant operations that happened. For example, creation or archival of ``Holdings``.
* Meta key/values, of which part of the standard:

  * ``splice.lfdecentralizedtrust.org/tx-kind``: the kind of operation happening in the node. This can give more information than the exercised choice does. It can be one of:

    * ``transfer``
    * ``merge-split``
    * ``burn``
    * ``mint``
    * ``unlock``
    * ``expire-dust``

  * ``splice.lfdecentralizedtrust.org/sender``: which party is the sender in the node.
  * ``splice.lfdecentralizedtrust.org/reason``: a text specifying the reason for the operation in the node.
  * ``splice.lfdecentralizedtrust.org/burned``: how much of a holding was burned in the node.

.. warning::

>     Meta key/values can be specified in several fields. For transfers, they should be merged in last-write-wins order of:

Reviewer: ```suggestion
    Meta key/values can be specified in several optional fields. For transfers, they values from fields that are present should be merged in last-write-wins order of:
```

    * event.choiceArgument?.transfer?.meta,
    * event.choiceArgument?.extraArgs?.meta,
    * event.choiceArgument?.meta,

>     * event.exerciseResult?.meta,

Reviewer: what do the `?` marks mean? Probably that these fields are not guaranteed to be present?

I'd remove them and use the suggestion I gave above to clarify that they are optional.

.. _token_standard_usage_executing_factory_choice:

Executing a factory choice
--------------------------

> `Token Standard CLI's code to create a transfer via TransferFactory <https://github.com/hyperledger-labs/splice/blob/main/token-standard/cli/src/commands/transfer.ts>`_

Reviewer: ```suggestion
Example code: `Token Standard CLI's code to create a transfer via TransferFactory <https://github.com/hyperledger-labs/splice/blob/main/token-standard/cli/src/commands/transfer.ts>`_
```

To execute a choice via a Token Standard factory, first you need need to fetch the factory from the corresponding registry.

.. note::

    The mapping from an instrument's `admin` party-id to the corresponding registry URL needs to be maintained currently by wallets themselves,

>     until a generic solution (likely based on CNS) is implemented.

Reviewer: ```suggestion
    until a generic solution ([likely based on CNS](https://github.com/global-synchronizer-foundation/cips/blob/main/cip-0056/cip-0056.md#off-ledger-api-discovery-and-access)) is implemented.
```

endpoints of the participant.
For non-external parties, you can just use the `submit-and-wait endpoint <https://github.com/digital-asset/canton/blob/f608ec2cbb7b3e9331b7cc564eb260916606d815/community/ledger/ledger-json-api/src/test/resources/json-api-docs/openapi.yaml#L1#L6>`_.

In both cases, you must include an ``ExerciseCommand`` in your payload with the following fields:

* ``templateId``: the interface id of the factory you want to exercise the choice on. For example, ``#splice-api-token-transfer-instruction-v1:Splice.Api.Token.TransferInstructionV1:TransferFactory``.
* ``contractId``: the ``factoryId`` obtained from the registry.
* ``choice``: the name of the choice you want to execute. For example, ``TransferFactory_Transfer``.
* ``choiceArgument``: the arguments that will be passed to the Daml choice. These will be decoded from JSON.
  For a ``TransferFactory_Transfer``, this will include for example the sender, receiver and amount, among other fields.

.. _token_standard_usage_executing_nonfactory_choice:

Executing a non-factory choice
------------------------------

> `Token Standard CLI's code to accept a transfer instruction <https://github.com/hyperledger-labs/splice/blob/main/token-standard/cli/src/commands/acceptTransferInstruction.ts>`_

Reviewer: ```suggestion
Example code: `Token Standard CLI's code to accept a transfer instruction <https://github.com/hyperledger-labs/splice/blob/main/token-standard/cli/src/commands/acceptTransferInstruction.ts>`_
```

* `To accept a TransferInstruction </app_dev/token_standard_openapi/index.html#post--registry-transfer-instruction-v1-transferInstructionId-choice-contexts-accept>`_
* `To reject a TransferInstruction </app_dev/token_standard_openapi/index.html#post--registry-transfer-instruction-v1-transferInstructionId-choice-contexts-reject>`_
* `To withdraw a TransferInstruction </app_dev/token_standard_openapi/index.html#post--registry-transfer-instruction-v1-transferInstructionId-choice-contexts-withdraw>`_
* `To withdraw an Allocation </app_dev/token_standard_openapi/index.html#post--registry-allocations-v1-allocationId-choice-contexts-withdraw>`_
* `To cancel an Allocation </app_dev/token_standard_openapi/index.html#post--registry-allocations-v1-allocationId-choice-contexts-cancel>`_

The response of these endpoints include two fields:

* ``choiceContextData``: to be passed as ``context`` in the ``choiceArgument``.
* ``disclosedContracts``: to be passed in the submit or prepare request.

.. warning::

  Note that ``AllocationRequest_Reject`` and ``AllocationRequest_Withdraw`` should be called with an empty choice context.

>   This is currently there as a potential future extension.

Reviewer: ```suggestion
  This `ChoiceContext` is present to allow for potential future extensions
  of the behavior of implementations of these choices.
```

..
   Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
..
   SPDX-License-Identifier: Apache-2.0

.. _token_standard_usage:

> Token Standard Usage

Reviewer: ```suggestion
Wallet Integration with Token Standard Assets
```

That seems to be what's being explained here.

====================

> See the `ledger's OpenAPI definition <https://github.com/hyperledger-labs/splice/blob/902f7069602b7bbce49293cabe7f162cfd8dbffb/canton/community/ledger/ledger-json-api/src/test/resources/json-api-docs/openapi.yaml#L1>`_.

Reviewer: This intro is missing an overview of the content that follows. Something along the lines of:

> This page provides wallet developers with guidance on how to integrate with token standard assets. Such an integration works by sending the right read and write requests to the Ledger API of the validator node hosting the wallet user's party. 
> 
> There are four kinds of integration patterns:
> <list the four> with links to the sections
> 
> All of these integration patterns are demonstrated in the form of executable code as part of the experimental command-line interface for token standard assets. The sections below explaining the patterns below thus all start with a link to the code. They then provide additional context for an implementor.  
> 

Follow this with a Sphinx note about where to find the JSON Ledger API docs, and the recommendation to use OpenAPI code generation tools.

Reviewer: These I'd consider workflows.

"Integration patterns" should probably be better named "interaction patterns": read ACS, read history, factory choice, non-factory choice.

> See the `ledger's OpenAPI definition <https://github.com/hyperledger-labs/splice/blob/902f7069602b7bbce49293cabe7f162cfd8dbffb/canton/community/ledger/ledger-json-api/src/test/resources/json-api-docs/openapi.yaml#L1>`_.

Reviewer: ```suggestion
See the OpenAPI definition of the `JSON Ledger API <https://github.com/digital-asset/canton/blob/main/community/ledger/ledger-json-api/src/test/resources/json-api-docs/openapi.yaml#L1>`_.
```

Reviewer: Please update the links to that asset. Permalinks might make sense.

> See the `ledger's OpenAPI definition <https://github.com/hyperledger-labs/splice/blob/902f7069602b7bbce49293cabe7f162cfd8dbffb/canton/community/ledger/ledger-json-api/src/test/resources/json-api-docs/openapi.yaml#L1>`_.

Reviewer: Yeah, not really. The One thing we can do better is to refer to the copy in the canton repo.

This is also accessible at ``http(s)://${YOUR_PARTICIPANT}/docs/openapi``.
We encourage developers to use OpenAPI code generation tools as opposed to manually writing HTTP requests.

Check out the :ref:`Authentication section <app-auth>` for more information on how to authenticate the requests.

.. _token_standard_usage_reading_contracts:

> Reading Contracts implementing a Token Standard interface for a party

Reviewer: ```suggestion
Reading contracts implementing a Token Standard interface for a party
```

---------------------------------------------------------------------

> `Token Standard CLI's code to list contracts by interface <https://github.com/hyperledger-labs/splice/blob/main/token-standard/cli/src/commands/listContractsByInterface.ts>`_

Reviewer: ```suggestion
Reference code from the Token Standard CLI  to `list contracts by interface <https://github.com/hyperledger-labs/splice/blob/main/token-standard/cli/src/commands/listContractsByInterface.ts>`_
```

The Token Standard includes several interfaces that are implemented by Daml templates.
To list all contracts implementing a particular interface,

> you have to query the participant's `active-contracts endpoint <https://github.com/hyperledger-labs/splice/blob/902f7069602b7bbce49293cabe7f162cfd8dbffb/canton/community/ledger/ledger-json-api/src/test/resources/json-api-docs/openapi.yaml#L620>`_.

Reviewer: I've asked in Slack whether there's a better option for these kinds of links: https://daholdings.slack.com/archives/C044LUG7CKT/p1754922860156509

Author: I don't think there's an answer in that thread?
I am now pointing to canton's as indicated [here](https://github.com/hyperledger-labs/splice/pull/1872#discussion_r2267027799)

                    "interfaceId": "$AN_INTERFACE_ID",
                    "includeInterfaceView": true,
                    "includeCreatedEventBlob": true
                  }
                }
              }
            }
          ]
        }
      }
    }

For example:

* ``"$A_PARTY"`` could look like ``test::1220a0db3761b3fc919b55e7ff80ad740824336010bfde8829611c0e64477ab7bee5``.
* ``"$AN_INTERFACE_ID"`` could be ``#splice-api-token-holding-v1:Splice.Api.Token.HoldingV1:Holding``.

To include other transaction nodes that don't directly involve the interfaces (e.g., non-interface-specific children nodes),
a ``WildcardFilter`` can be included in the ``cumulative`` filter array:

.. code-block:: json

    {
      "identifierFilter": {
        "WildcardFilter": {
          "value": {
            "includeCreatedEventBlob": true
          }
        }
      }
    }

The ``beginExclusive`` field is the offset from which to start reading transactions.
To paginate, you can start with the ``participantPrunedUpToInclusive`` from ``GET ${PARTICIPANT_URL}/v2/state/latest-pruned-offsets``
and continue by passing the offset of the last transaction from the previous response.

> Parsing
> ^^^^^^^

Reviewer: ```suggestion
Parsing the history
^^^^^^^^^^^^^^^^^^^
```

    * ``transfer``
    * ``merge-split``
    * ``burn``
    * ``mint``
    * ``unlock``
    * ``expire-dust``

  * ``splice.lfdecentralizedtrust.org/sender``: which party is the sender in the node.
  * ``splice.lfdecentralizedtrust.org/reason``: a text specifying the reason for the operation in the node.
  * ``splice.lfdecentralizedtrust.org/burned``: how much of a holding was burned in the node.

.. warning::

    Meta key/values can be specified in several fields. For transfers, they should be merged in last-write-wins order of:

    * event.choiceArgument?.transfer?.meta,
    * event.choiceArgument?.extraArgs?.meta,
    * event.choiceArgument?.meta,
    * event.exerciseResult?.meta,

.. _token_standard_usage_executing_factory_choice:

Executing a factory choice
--------------------------

`Token Standard CLI's code to create a transfer via TransferFactory <https://github.com/hyperledger-labs/splice/blob/main/token-standard/cli/src/commands/transfer.ts>`_

To execute a choice via a Token Standard factory, first you need need to fetch the factory from the corresponding registry.

> .. warning::

Reviewer: ```suggestion
.. note::
```

>     Getting the corresponding registry needs to be maintained currently by apps, until a generic solution is implemented.

Reviewer: ```suggestion
    The mapping from an instrument's `admin` party-id to the corresponding registry URL needs to be maintained currently by wallets themselves, until a generic solution (likely based on CNS) is implemented.
```

* `To reject a TransferInstruction </app_dev/token_standard_openapi/index.html#post--registry-transfer-instruction-v1-transferInstructionId-choice-contexts-reject>`_
* `To withdraw a TransferInstruction </app_dev/token_standard_openapi/index.html#post--registry-transfer-instruction-v1-transferInstructionId-choice-contexts-withdraw>`_
* `To withdraw an Allocation </app_dev/token_standard_openapi/index.html#post--registry-allocations-v1-allocationId-choice-contexts-withdraw>`_
* `To cancel an Allocation </app_dev/token_standard_openapi/index.html#post--registry-allocations-v1-allocationId-choice-contexts-cancel>`_

The response of these endpoints include two fields:

* ``choiceContextData``: to be passed as ``context`` in the ``choiceArgument``.
* ``disclosedContracts``: to be passed in the submit or prepare request.

.. warning::

  Note that ``AllocationRequest_Reject`` and ``AllocationRequest_Withdraw`` should be called with an empty choice context.
  This is currently there as a potential future extension.

.. _token_standard_usage_workflows:

Token Standard Workflows

> ------------------------

Reviewer: The sections below lack context as to why these are relevant and what's their difference.

I'd suggest that you reference the corresponding sections in the standard:
- https://github.com/global-synchronizer-foundation/cips/blob/main/cip-0056/cip-0056.md#direct-peer-to-peer--free-of-payment-fop-transfer-workflow
- https://github.com/global-synchronizer-foundation/cips/blob/main/cip-0056/cip-0056.md#delivery-versus-payment-dvp-transfer-workflows

And inline a copy of them interleaved with the technical content that you reference below.

I'd also consider moving this section up towards the top, as it gives context for the different kinds ledger interactions explained in the above sections.

wdyt?

* Depending on the implementation of the factory, one of three ``TransferInstructionResult_Output`` is possible:

  * ``Failed``: where the transfer did not succeed and all holdings (minus fees) have been returned to the sender.
  * ``Completed``: where the transfer succeeded and the receiver has received their holdings. No further action is required.
  * ``Pending``: where the transfer is pending further steps. This will include a ``transferInstructionCid``.
* If ``Pending``, The receiver party observes a ``TransferInstruction`` (which has the same contract id as above). Then:

  * The receiver can exercise ``TransferInstruction_Accept``, which again will return a ``TransferInstructionResult`` depending on success and whether further steps are required or not.
  * The receiver can exercise ``TransferInstruction_Reject``, same as above.
  * The sender can exercise ``TransferInstruction_Withdraw``, again returning a ``TransferInstructionResult``.
  * The registry can exercise ``TransferInstruction_Update``, again returning a ``TransferInstructionResult``.

Allocations
^^^^^^^^^^^

The workflow will follow the following steps:

> * A registry creates as many ``AllocationInstructions`` as required for a workflow to happen.

Reviewer: This seems wrong. What did you mean here? this sounds like you wanted to talk about `AllocationRequest`s?

* Parties can:

  * The registry can exercise ``AllocationInstruction_Withdraw`` or ``AllocationInstruction_Update`` on it.

>   * The senders of each transfer leg can exercise ``AllocationFactory_Allocate`` to create an ``Allocation`` satisfying the conditions of the ``AllocationInstruction``.

Reviewer: ```suggestion
  * The senders of each transfer leg can exercise ``AllocationFactory_Allocate`` to create an ``Allocation`` satisfying the conditions of the ``AllocationRequest``.
```

>   * The following choices can be called on the ``Allocation``:

Reviewer: Consider referencing the source code and its commentary. So that people become accustomed to actually reading it and using it as guidance.

>     * Sender, receiver and registry can exercise ``Allocation_ExecuteTransfer``: to accept the allocation.

Reviewer: ```suggestion
    * Sender, receiver and registry can jointly exercise ``Allocation_ExecuteTransfer``: to execute the allocated transfer.
```
